File: generic-code-challenges/largest-sum/largest_sum.py
def subarray_greatest_sum(numbers):
    if len(numbers) < 2:
        return numbers

    max_sum = numbers[0]
    max_i = 0
    max_j = 0

    for idx, item in enumerate(numbers):
        current_sum = 0
        for j in range(idx, len(numbers)):
            current_sum = sum(numbers[idx: j])
            if current_sum > max_sum:
                max_sum = current_sum
                max_i = idx
                max_j = j

    return numbers[max_i:max_j]


def kadanes_algorithm(numbers):
    """Solution in O(N) time."""
    max_here = max_so_far = numbers[0]
    start = end = 0
    for i in range(1, len(numbers)):
        if numbers[i] > max_here + numbers[i]:
            max_here = numbers[i]
            start = end = i
        else:
            max_here += numbers[i]
            if max_here > max_so_far:
                end = i + 1

        if max_here > max_so_far:
            max_so_far = max_here

    return numbers[start: end]

File: generic-code-challenges/largest-sum/largest-sum.js
const biggest_sum = (numbers) => {
  // Simple edge case
  if ( numbers.length < 2 ) { return numbers; }

  // Starting point is zero and between elements 0 and 0 (i.e. the first element)
  let max_sum = numbers[0];
  let max_start = 0;
  let max_end = 0;


  numbers.forEach( (num,idx) => {

    // New sum set for each starting point (idx) looking ahead
    let current_sum = 0;

    // Start where we are (idx) and look ahead
    for( let j = idx; j<= numbers.length; j++ ) { 
      current_sum += numbers[j]; // Increment the sum

      if ( current_sum > max_sum ) { 
        max_sum = current_sum;  // update the best sum so far if we've beaten it
        max_start = idx;        // The slice will now start at the outer index (idx)
        max_end = j;            // The slice will end here (j)
      }
    }

  });

  return numbers.slice(max_start, max_end+1);

}


const nums = [-68, -76, 55, -60, 82, -12, -95, 72, 87, 63, -53, -93, 68, -10, -4, 86, 18, -40, 19, -34]

console.log( biggest_sum(nums) );

// [72, 87, 63, -53, -93, 68, -10, -4, 86, 18]
